modules/openapi/src/internals/postprocess/AllOfTransformer.scala

```diff
@@ -18,86 +18,215 @@ package postprocess
+      newDef: Structure
+  ): TopLevelParentNoRefsResult = {
+    // 1. Make structure parent a mixin
+    val newParent = parent.copy(hints = parent.hints :+ Hint.IsMixin)
+    // 2. Apply as mixin to this structure instead of moving all fields in
+    val nd = newDef.copy(hints = newDef.hints :+ Hint.HasMixin(parent.id))
+    TopLevelParentNoRefsResult(nd, newParent)
+  }
 
-  @tailrec
-  private def getAllParentFields(
-      newId: DefId,
-      allShapes: Vector[Definition],
-      parents: Vector[Definition],
-      state: State = State.empty
-  ): State = parents.toList match {
-    case parent :: tail =>
-      parent match {
-        case Structure(id, local, newParents, hints) =>
-          val updatedLocal =
-            local.map(f => f.copy(id = f.id.copy(modelId = newId)))
-          val newState = state.addFields(updatedLocal).withHints(hints)
-          val finalState =
-            if (!hints.contains(Hint.TopLevel)) newState.addRemoveDef(id)
-            else newState
-          getAllParentFields(
-            newId,
-            allShapes,
-            tail.toVector ++ allShapes.filter(s => newParents.contains(s.id)),
-            finalState
-          )
-        case Newtype(_, DocumentPrimitive, hints) =>
-          val newState = state.setIsDocument.withHints(hints)
-          getAllParentFields(newId, allShapes, tail.toVector, newState)
-        case _ => getAllParentFields(newId, allShapes, tail.toVector, state)
-      }
-    case Nil => state
+  private case class NonTopLevelParentNoRefsResult(
+      newDef: Structure,
+      removeParent: Structure
+  )
+  private def nonTopLevelParentNoReferences(
+      parent: Structure,
+      newDef: Structure
+  ): NonTopLevelParentNoRefsResult = {
+    // Move fields down from parent and remove parent if parent is not a top level
+    // shape (meaning it is a fabricated AllOf shape created by the OpenApi => IModel transform)
+    val newFields =
+      parent.localFields.map(f => f.copy(id = f.id.copy(modelId = newDef.id)))
+    val remove = parent
+    val nd = newDef.copy(localFields = newDef.localFields ++ newFields)
+    NonTopLevelParentNoRefsResult(nd, remove)
   }
 
-  private def transform(in: IModel): Vector[Definition] = {
-    val allShapes = in.definitions
-    val (remove, result) = allShapes
-      .map {
-        case s @ Structure(id, localFields, parents, _) if parents.nonEmpty =>
-          val parentDefs =
-            allShapes.filter(definition => parents.contains(definition.id))
-          val State(fieldsToAdd, defsToRemove, isDocument, hints) =
-            getAllParentFields(id, allShapes, parentDefs)
-          val newStruct = s.copy(
-            localFields = localFields ++ fieldsToAdd,
-            parents = Vector.empty,
-            hints = s.hints ++ hints
+  private def moveParentFieldsAndCreateMixins(
+      all: Vector[Definition]
+  ): Vector[Definition] = {
+    val allShapes: ListBuffer[Definition] = ListBuffer.from(all)
```

Comment:
could we use a `mutable.Map[ShapeId, Definition]` here

you'd have `add/remove/update(replace)` out of the box?

Reply:
Yeah definitely a good idea 👍

Reply:
Handled this in latest two commits, needed to use `LinkedHashMap` as ordering of the shapes should be preserved for things like union transforms that occur after this transform.